src/ngraph/runtime/dynamic/dynamic_backend.cpp

```diff
@@ -117,130 +117,210 @@ bool runtime::dynamic::DynamicExecutable::call(
     // (1) all shapes;
     // (2) all values of shape-relevant input tensors.
 
-    NGRAPH_CHECK(m_wrapped_function->get_parameters().size() == inputs.size());
-
-    std::vector<std::shared_ptr<runtime::Tensor>> wrapped_inputs;
-    std::vector<element::Type> arg_element_types;
-    std::vector<PartialShape> arg_shapes;
-
-    std::shared_ptr<Function> clone;
+    std::vector<int> merged_input_shapes;
+    std::ostringstream key;
+    size_t loop_count = 0;
+    for (auto& input : inputs)
     {
-        // We'll use AlignedBuffers to back the base pointers, storing them in this vector for RAII
-        // purposes.
-        std::vector<AlignedBuffer> arg_buffers;
-        arg_buffers.reserve(inputs.size());
-        std::vector<void*> arg_value_base_pointers(inputs.size());
-
-        size_t i = 0;
-
-        for (auto& input : inputs)
+        if (m_wrapped_function->get_parameters()[loop_count]->is_relevant_to_shapes())
         {
-            if (m_wrapped_function->get_parameters()[i]->is_relevant_to_shapes())
+            // Caching on values of Shape relevant inputs
+            void* data_ptr = malloc(input->get_size_in_bytes());
+            input->read(data_ptr, input->get_size_in_bytes());
+            // assuming that shape relevant inputs are int64_t
+            int* result = (int*)data_ptr;
+            for (int i = 0; i < input->get_element_count(); i++)
             {
-                // TODO(amprocte): Move has_storage() to runtime::Tensor?
-                if (auto dynamic_tensor =
-                        std::dynamic_pointer_cast<runtime::dynamic::DynamicTensor>(input))
-                {
-                    NGRAPH_CHECK(dynamic_tensor->has_storage());
-                }
-
-                arg_buffers.emplace_back(input->get_size_in_bytes(), /*alignment=*/64);
-                arg_value_base_pointers[i] = arg_buffers.back().get_ptr();
-
-                // TODO(amprocte): For host-resident tensors we should be able to skip the read,
-                // but no API for that yet.
-                input->read(arg_value_base_pointers[i], input->get_size_in_bytes());
+                merged_input_shapes.emplace_back(result[i]);
             }
-            else
+        }
+        else
+        {
+            // Caching on all remaining shapes
+            for (int i = 0; i < input->get_shape().size(); i++)
             {
-                arg_value_base_pointers[i] = nullptr;
+                merged_input_shapes.emplace_back(input->get_shape()[i]);
             }
+        }
+        // -1 is the separator.
+        // So if shape of Input 1 = {2, 2, 3, 3} & Input 2 = {4, 5}
+        // the key would be 2, 2, 3, 3, -1, 4, 5, -1
+        merged_input_shapes.emplace_back(-1);
+        loop_count++;
+    }
```

Comment:
Minor:
This computation of the key could be moved inside the cache class. then the calling code does not have to do this. The class could just accept a `vector<vector<int>>` or `vector<ng::Shape>` or something else and compute its key as it sees fit. The user does not have to know about how the class implements the key calculation